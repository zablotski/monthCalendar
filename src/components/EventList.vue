<template>
  <ul class="event-list">
    <li v-for="event of day.events" :key="event" class="event-item">
      <span class="event-text">{{ event.text }}</span>
      <span class="actions-list">
        <button @click.stop="() => deleteEvent(day, event)">X</button>
        <button @click.stop="() => editEvent(day, event)">&#9998;</button>
      </span>
    </li>
  </ul>
</template>
<script>
export default {
  name: "EventList",
  props: {
    day: { events: Array },
    editEvent: Function,
    deleteEvent: Function,
  },
};
</script>
<style scoped>
ul.event-list {
  display: flex;
  flex-direction: column;
  padding: 0;
  width: 100%;
}

ul.event-list {
  display: flex;
  flex-direction: column;
}

.event-item {
  font-size: 1rem;
  list-style: none;
  background-color: #fff;
  border: 1px solid lightgrey;
  padding: 0.125rem;
  display: grid;
  grid-template-columns: repeat(2, auto);
}

.actions-list {
  visibility: hidden;
  display: flex;
  justify-self: end;
}

.event-item:hover .actions-list {
  visibility: visible;
}

.actions-list > * {
  margin-left: 0.125rem;
}

.event-text {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}
</style>
