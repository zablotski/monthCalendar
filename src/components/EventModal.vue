<script>
export default {
  methods: {
    onWrapperClick(e) {
      if (e.target === this.$refs.modalWrapper) {
        this.$emit("close");
      }
    },
  },
  emits: ["close"],
};
</script>

<template>
  <div class="modal-mask">
    <div class="modal-wrapper" @click="onWrapperClick" ref="modalWrapper">
      <div class="modal-container">
        <div class="modal-header">
          <slot name="header"> </slot>
          <button class="modal-default-button" @click="$emit('close')">
            X
          </button>
        </div>

        <div class="modal-body">
          <slot name="body"></slot>
        </div>

        <div class="modal-footer">
          <slot name="footer"> </slot>
        </div>
      </div>
    </div>
  </div>
</template>

<style>
.modal-mask {
  position: fixed;
  z-index: 9998;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: table;
}

.modal-wrapper {
  display: table-cell;
  vertical-align: middle;
}

.modal-container {
  width: 300px;
  margin: 0 auto;
  padding: 1rem;
  background-color: #fff;
}

.modal-header h3 {
  margin-top: 0;
  color: lightgrey;
}

.modal-body {
  margin: 1.25rem 0;
}

.modal-default-button {
  float: right;
}
</style>
