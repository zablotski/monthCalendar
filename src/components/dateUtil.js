export const getDates = (start, end) => {
  let result = [];

  for (
    let date = new Date(start);
    date <= new Date(end);
    date.setDate(date.getDate() + 1)
  ) {
    result.push(new Date(date));
  }
  return result;
};

const getFirstDayOfWeek = (date, firstDayOfWeek = 1) => {
  const day = date.getDay();
  const diff = (day < firstDayOfWeek ? 7 : 0) + day - firstDayOfWeek;

  date.setDate(date.getDate() - diff);
  date.setHours(0, 0, 0, 0);
  return date;
};

const getLastDayOfWeek = (date, firstDayOfWeek = 1) => {
  const day = date.getDay();
  const diff = (day < firstDayOfWeek ? -7 : 0) + 6 - (day - firstDayOfWeek);

  date.setDate(date.getDate() + diff);
  date.setHours(23, 59, 59, 999);

  return date;
};

export const getCalendarDates = (year, month) => {
  const firstDayOfMonth = new Date(year, month, 1);
  const lastDayOfMonth = new Date(
    firstDayOfMonth.getFullYear(),
    firstDayOfMonth.getMonth() + 1,
    0
  );

  return getDates(
    getFirstDayOfWeek(firstDayOfMonth),
    getLastDayOfWeek(lastDayOfMonth)
  );
};
