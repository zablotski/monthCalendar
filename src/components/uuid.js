export const randomString = () => Math.random().toString(36).replace("0.", "");
