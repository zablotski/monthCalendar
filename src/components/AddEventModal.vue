<template>
  <EventModal @close="$emit('close')">
    <template #header>Add Event:</template>
    <template #body>
      <input type="text" v-model="text" name="newEvent" id="newEventInput" />
      <button class="modal-default-button" @click="$emit('submit', text)">
        OK
      </button>
    </template>
  </EventModal>
</template>
<script>
import EventModal from "@/components/EventModal.vue";

export default {
  name: "AddEventModal",
  data: function () {
    return {
      text: "",
    };
  },
  components: { EventModal },
  emits: ["submit", "close"],
};
</script>
<style scoped></style>
