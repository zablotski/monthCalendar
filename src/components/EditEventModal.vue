<template>
  <EventModal @close="$emit('close')">
    <template #header>Edit Event:</template>
    <template #body>
      <input type="text" v-model="text" name="editEvent" id="editEventInput" />
      <button class="modal-default-button" @click="$emit('submit', text)">
        OK
      </button>
    </template>
  </EventModal>
</template>
<script>
import EventModal from "@/components/EventModal.vue";
export default {
  components: { EventModal },
  name: "EditEventModal",
  data() {
    return {
      text: this.event.text,
    };
  },
  props: {
    event: Object,
    day: Object,
  },
};
</script>
<style scoped></style>
