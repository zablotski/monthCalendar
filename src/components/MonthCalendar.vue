<script>
import { getCalendarDates } from "@/components/dateUtil";
import AddEventModal from "@/components/AddEventModal.vue";
import EditEventModal from "@/components/EditEventModal.vue";
import { randomString } from "@/components/uuid";
import EventList from "@/components/EventList.vue";

export default {
  components: { EventList, EditEventModal, AddEventModal },
  data() {
    return {
      date: null,
      calendarData: [],
      showAddModal: false,
      showEditModal: false,
      showRemoveModal: false,
      selectedDay: null,
      selectedEvent: null,
    };
  },
  computed: {},
  created() {
    this.date = new Date();

    const calendarDates = getCalendarDates(
      this.date.getFullYear(),
      this.date.getMonth()
    );

    this.mapDatesToData(calendarDates);
  },
  methods: {
    onDayClick(day) {
      this.showAddModal = true;
      this.selectedDay = day;
    },
    addEvent(text) {
      const trimmed = text.trim();
      if (!text) {
        return;
      }
      this.selectedDay.events.push({
        id: randomString(),
        text: trimmed,
      });
      this.showAddModal = false;
    },
    editEvent(day, event) {
      this.selectedDay = day;
      this.selectedEvent = event;
      this.showEditModal = true;
    },
    deleteEvent(day, event) {
      day.events = day.events.filter((e) => e !== event);
    },
    mapDatesToData: function (calendarDates) {
      this.calendarData = calendarDates.map((date) => ({
        dateObject: date,
        events: [],
      }));
    },
  },
};
</script>

<template>
  <div class="calendar">
    <header>
      <h1>
        {{ this.date.toLocaleString("default", { month: "long" }) }}
        {{ this.date.getFullYear() }}
      </h1>
    </header>

    <ul class="grid weekdays">
      <li v-for="day of calendarData.slice(0, 7)" :key="day">
        {{ day.dateObject.toLocaleString("default", { weekday: "long" }) }}
      </li>
    </ul>

    <ol class="grid day-grid">
      <li
        v-for="day of calendarData"
        :key="day"
        :class="{
          'other-month': day.dateObject.getMonth() !== this.date.getMonth(),
          day: true,
        }"
        @click.exact="() => onDayClick(day)"
      >
        {{ day.dateObject.getDate() }}
        <EventList
          :day="day"
          :edit-event="editEvent"
          :delete-event="deleteEvent"
        />
      </li>
    </ol>
    <AddEventModal
      v-if="showAddModal"
      @close="showAddModal = false"
      @submit="addEvent"
    />
    <EditEventModal
      v-if="showEditModal"
      @close="showEditModal = false"
      :event="selectedEvent"
      :day="selectedDay"
      @submit="
        (text) => {
          selectedEvent.text = text;
          showEditModal = false;
        }
      "
    />
  </div>
</template>
<style scoped>
.calendar {
  font-family: system-ui;
}

header {
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 1em;
  background: #000;
  color: #fff;
  min-height: 5vh;
  text-align: center;
}

ul.grid,
ol.grid {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-gap: 1em;
  margin: 0 auto;
  max-width: 64em;
  padding: 0;
}

li {
  display: flex;
  align-items: center;
  justify-content: center;
  list-style: none;
  margin-left: 0;
  font-size: 1rem;
  cursor: pointer;
  flex-direction: column;
}

li.day {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

ul.event-list li {
  font-size: 0.75rem;
  list-style: none;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

ul.weekdays {
  margin-bottom: 1em;
}

ul.weekdays li {
  height: 2vw;
}

ol.day-grid > li {
  background-color: lightgrey;
  border: 1px solid lightgrey;
  height: 12vw;
  max-height: 125px;
  overflow: hidden;
}

ol.day-grid li.other-month {
  background-color: #fff;
}

@media all and (max-width: 800px) {
  ul,
  ol {
    grid-gap: 0.25em;
  }
}
</style>
