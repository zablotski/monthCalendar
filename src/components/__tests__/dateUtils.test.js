import { describe, it, expect } from "vitest";
import { getCalendarDates } from "@/components/dateUtil";

describe("dateUtils: ", () => {
  it("should return calendar data", () => {
    const result = getCalendarDates(2022, 8);
    expect(result).toMatchSnapshot();
  });
});
