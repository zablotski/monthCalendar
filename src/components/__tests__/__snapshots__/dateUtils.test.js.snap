// Vitest Snapshot v1

exports[`dateUtils:  > should return calendar data 1`] = `
[
  2022-08-28T22:00:00.000Z,
  2022-08-29T22:00:00.000Z,
  2022-08-30T22:00:00.000Z,
  2022-08-31T22:00:00.000Z,
  2022-09-01T22:00:00.000Z,
  2022-09-02T22:00:00.000Z,
  2022-09-03T22:00:00.000Z,
  2022-09-04T22:00:00.000Z,
  2022-09-05T22:00:00.000Z,
  2022-09-06T22:00:00.000Z,
  2022-09-07T22:00:00.000Z,
  2022-09-08T22:00:00.000Z,
  2022-09-09T22:00:00.000Z,
  2022-09-10T22:00:00.000Z,
  2022-09-11T22:00:00.000Z,
  2022-09-12T22:00:00.000Z,
  2022-09-13T22:00:00.000Z,
  2022-09-14T22:00:00.000Z,
  2022-09-15T22:00:00.000Z,
  2022-09-16T22:00:00.000Z,
  2022-09-17T22:00:00.000Z,
  2022-09-18T22:00:00.000Z,
  2022-09-19T22:00:00.000Z,
  2022-09-20T22:00:00.000Z,
  2022-09-21T22:00:00.000Z,
  2022-09-22T22:00:00.000Z,
  2022-09-23T22:00:00.000Z,
  2022-09-24T22:00:00.000Z,
  2022-09-25T22:00:00.000Z,
  2022-09-26T22:00:00.000Z,
  2022-09-27T22:00:00.000Z,
  2022-09-28T22:00:00.000Z,
  2022-09-29T22:00:00.000Z,
  2022-09-30T22:00:00.000Z,
  2022-10-01T22:00:00.000Z,
]
`;
